pandas/tests/frame/test_reshape.py

```diff
@@ -1002,6 +1002,26 @@ def test_stack_preserve_categorical_dtype_values(self):
         )
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize(
+        "index, columns",
+        [
+            ([0, 0, 1, 1], pd.MultiIndex.from_product([[1, 2], ["a", "b"]])),
+            ([0, 0, 2, 3], pd.MultiIndex.from_product([[1, 2], ["a", "b"]])),
+            ([0, 1, 2, 3], pd.MultiIndex.from_product([[1, 2], ["a", "b"]])),
+        ],
+    )
+    def test_stack_multi_columns_non_unique_index(self, index, columns):
+        # GH-28301
+        df = pd.DataFrame(index=index, columns=columns).fillna(1)
+        stacked = df.stack(-1)
+        new_index = pd.MultiIndex.from_tuples(stacked.index.values)
+
+        tm.assert_index_equal(stacked.index, new_index)
+
+        a = np.asarray(stacked.index.codes)
+        b = np.asarray(new_index.codes)
```

Comment:
can u created the full expected frame here and compare that instead

Reply:
Added in frame comparison, I left the `assert_numpy_equal` because it doesn't seem like `assert_frame_equal` checks that the codes on each of the indexes match.